// modes/modeManager.js
var prelude = require('prelude-ls');

var oop = require('std/oop');

var SessionManager = require('session').SessionManager;
var Code = require('code').Code;







var ModeInput = oop.Class({
	name: "Mode Input",

	desc: "This represents the input data used to add a mode to the mode manager. It is simply an association of one or multiple names to an instance of a mode.",

	schema:  {
		properties: [
			{names: ['names', 'name', 'keys', 'key', 'ids', 'id'], ctor: oop.types.Array, required: true},
			{names: ['mode', 'module'], required: true}
		]
	}
});



var SessionInit = oop.Class({
	name: "Session Init",

	desc: "This represents the input data used to init a document. It is mandatory to tell in which mode/language your document will work, and you can optionally provide an initial content.",

	schema: {
		properties: [
			{names: ['mode', 'lang', 'language'], required: true},
			{names: ['source', 'src', 'text', 'input', 'content']}
		]
	}
});


var idPropertyNames = ['id', 'ID', 'guid', 'GUID', 'key', 'session'];

var Request = oop.Class({
	name: "Mode Request",

	desc: "A mode request is the input for querying a service from a mode",

	schema: {
		properties: [
			{names: idPropertyNames, required: true},
			{names: ['service', 'svc', 'method', 'member'], required: true},
			{names: ['arg', 'args', 'argument', 'arguments'], default: {}}
		]
	}
});

var DocumentID = oop.Class({
	name: "Document ID",

	desc: "The GUID/ID object used to identify a document",

	schema: {
		inputToSpec: {
			'String': 'id'
		},
		properties: [
			{names: idPropertyNames, type: oop.types.Number, required: true}
		]
	}
});


var ModeManager = new oop.Class({
	name: "Mode Manager",

	constructor: function(input) {
		this.modes = {};
		this.sessions = new SessionManager();
	},

	prototype: {
		source: function(input) {
			var documentID = DocumentID.factory(input);
			return this.getSessionData(documentID.id).code.source;
		},

		/***********************************************************************
		 * Modes management
		 **********************************************************************/

		add: function(input) {
			if (input == null) {
				throw {
					msg: 'Missing input for method'
				}
			}

			var modeInputs = oop.types.Array.factory(input);

			for (var i = 0; i < modeInputs.length; i++) {
				var modeInput = ModeInput.factory(modeInputs[i]);

				var names = modeInput.names;
				for (var j = 0; j < names.length; j++) {
					this.modes[names[j]] = modeInput.mode;
				}
			}
		},

		exec: function(input) {
			var request = Request.factory(input);
			if (request == null) {
				throw {
					msg: 'Missing input'
				}
			}

			var data = this.getSessionData(request.id);

			var mode = this.modes[data.mode];
			if (mode == null) {
				throw {
					msg: 'Mode not available',
					mode: data.mode
				}
			}

			var service = mode[request.service];
			if (service == null) {
				throw {
					msg: 'Service not available',
					service: request.service,
					request: request,
					input: input,
					mode: data.mode
				}
			}

			try {
				return service.call(mode, data.code, request.arg);
			} catch (exception) {
				if (exception instanceof Error) {
					exception = exception.stack;
				}

				throw {
					msg: 'A problem occured while executing a service',
					service: request.service,
					mode: data.mode,
					request: request,
					exception: exception
				}
			}
		},

		/***********************************************************************
		 * General editor services
		 **********************************************************************/

		configuration: function() {
			return  {
				tabWidth: 4
			}
		},

		/***********************************************************************
		 * Session management
		 **********************************************************************/

		/**
		 * Inits a new session.
		 *
		 * A session is used to manage an instance of a source code on the client side: basically a whole program, a file.
		 *
		 * @todo Maybe allow to pass initial data, like the source
		 */
		init: function(input) {
			if (input == null) {
				throw {
					msg: 'Missing input for method'
				}
			}
			var spec = SessionInit.factory(input);

			// mode ------------------------------------------------------------

			var mode = spec.mode;

			if (!(mode in this.modes)) {
				throw {
					msg: 'Unavailable mode',
					mode: mode,
					modes: this.modes
				}
			}

			// source ----------------------------------------------------------

			var code = this.modes[mode].create(spec.source);

			// Initialization --------------------------------------------------

			var id = this.sessions.init({
				code: code,
				mode: mode
			});

			// Response --------------------------------------------------------

			return {
				guid: id
			};
		},

		getSessionData: function(id) {
			if (id == null) {
				throw {
					msg: "No id given"
				}
			}

			var data = this.sessions.get(id);

			if (data == null) {
				throw {
					msg: 'Invalid ID (no data)',
					id: id
				};
			}

			return data;
		}
	}
});



exports.ModeManager = ModeManager;
