// modes/node_modules/code.js
var oop = require('oop');

var prelude = require('prelude-ls');


/**
 * Model for a source code.
 *
 * Kind of equivalent (but more specific) to the Document model in Eclipse.
 */
var Code = oop.class({
	name: 'Code',

	constructor: function(parser, input) {
		this.parser = parser;
		this.source = '';
		this.graph = undefined;
		if (input != null) {
			this.update(input);
		}
	},

	methods: {
		/**
		 * WARNING: Be careful of being efficient in this method, it is critical as this is the one called the most often!
		 *
		 * @todo? be able to specify a length property (computing end or start)
		 * @todo? be able to specify a unique end property for insertion (like it does for now with unique start)
		 * @todo be able to specify a pos property, for insertion, less confusing that the unique start or end property
		 */
		update: {
			chainable: true,

			input: {
				inputToSpec: {
					'String': 'text'
				},
				properties: [
					{
						names: ['text', 'source', 'src', 'content'],
						default: '',
						type: oop.types.String
					},
					{
						names: ['start', 'beginning', 'offset', 'index'],
						default: function() {return this.source.length;}, // appended text by default
						type: oop.types.Number
					},
					{
						names: ['end', 'stop'],
						// inferred: 'start'
						type: oop.types.Number
					},
					{
						names: ['replace', 'erase', 'swap'],
						default: false,
						type: oop.types.Boolean
					},
				]
			},

			exec: function(spec) {
				if (spec.replace) {
					this.source = spec.text;
					this.graph = this.parser.parse(this.source);
				} else {
					if (spec.end == null) {
						spec.end = spec.start;
					}

					this.source =
						this.source.substring(0, spec.start) +
						spec.text +
						this.source.substring(spec.end, this.source.length - 1);

					this.graph = this.parser.parse(this.source);
				}
			}
		}
	}
})



exports.Code = Code;
