// highlight/font.js
var oop = require('oop');



var Font = oop.Class({
	name: 'Font',

	desc: 'Represents font data.',

	schema: {
		inputToSpec: {
			'String': 'family',
			'Number': 'height',
			'Boolean': 'bold'
		},
		properties: [
			{names: ['family', 'f'], type: oop.types.String},
			{names: ['height', 'h', 'size'], type: oop.types.Number},
			{names: ['bold', 'b'], type: oop.types.Boolean},
			{names: ['italic', 'i'], type: oop.types.Boolean},
			{names: ['strike', 's', 'lineThrough', 'line-through'], type: oop.types.Boolean},
			{names: ['underline', 'u'], type: oop.types.Boolean}
		]
	},

	methods: {
		css: function() {
			var declarations = [];
			if (this.family != null) declarations.push("font-family:'" + this.family + "'");
			if (this.height != null) declarations.push("font-size:" + this.height + "px");
			if (this.bold) declarations.push("font-weight:bold");
			if (this.italic) declarations.push("font-style:italic");
			if (this.underline || this.strike) {
				var attributes = [];
				if (this.underline) attributes.push("underline");
				if (this.strike) attributes.push("line-through");
				declarations.push("text-decoration:" + attributes.join(' '));
			}
			return declarations;
		}
	},

	factories: {
		standard: {
			values: {
				family: 'Arial',
				height: '12',
				bold: false,
				italic: false,
				strike: false,
				underline: false
			}
		},
		family: {
			args: {family: 0},
			values: {
				height: '12',
				bold: false,
				italic: false,
				strike: false,
				underline: false
			}
		},

	}
});



exports.Font = Font;
