// highlight/stylesheet.js
var prelude = require('prelude-ls');

var oop = require('oop').oop;

var Style = require('./style').Style;



var Stylesheet = oop.Class({
	name: 'Stylesheet',

	desc: 'A stylesheet is a set of styles, each of them associated to a unique id. A stylesheet also contains a specific style which is the default style: if ever you try to get the style for an id which doesn\'t exist, the default style will be returned.',

	constructor: function() {
		this.styles = {};
		this._styleRegExps = {};
		this._sortedSelectors = null;
		this.default = undefined;

	},
	statics: {
		CLASS_PREFIX: "editor-highlight-",
		DEFAULT_CLASS: "token"
	},
	methods: {
		add: {
			input: {
				properties: [
					{name: 'style', type: Style, mixed: true},
					{names: ['selector', 'id', 'key']}
				]
			},

			process: function(spec) {
				if (spec.selector == null) {
					this['default'] = spec.style;
				} else {
					this.styles[spec.selector] = spec.style;
					this._styleRegExps[spec.selector] = new RegExp("(\\.|^)" + spec.selector.replace(/\./g, "\\.") + "$");
				}

				this._sortedSelectors = Object.keys(this.styles).sort(function (a, b) {
					var revA = a.split("").reverse().join("");
					var revB = b.split("").reverse().join("");
					if (revA > revB) {
						return -1;
					}
					if (revA == revB) {
						return 0;
					}
					return 1;
				});
			},

			output: 'style'
		},


		/**
		 * Convert a selector into a valid classname
		 */
		convertSelectorToClassName: function(selector) {
			return selector.replace(/\./g, "-");
		},

		/**
		 * Returns the classname to apply for a certain fullpath
		 */
		 getStyleFromPath : function(fullpath) {
		 	var selector;
		 	for (var i = 0, length = this._sortedSelectors.length; i < length; i++) {
		 		selector = this._sortedSelectors[i];
		 		if (fullpath.match(this._styleRegExps[selector])) {
		 			return selector;
		 		}
		 	}
		 	return null;
		 },

		/***********************************************************************
		 * Serialization
		 **********************************************************************/

		/**
		 * @todo Use the class $token as a master style
		 */
		css: function() {
			var ruleSets = [];
			ruleSets.push("." +  Stylesheet.CLASS_PREFIX + Stylesheet.DEFAULT_CLASS + "{" + this.default.css().join(';') + "}");
			for (var selector in this.styles) {
				var style = this.styles[selector];
				var declarations = style.css();
				ruleSets.push('.' + Stylesheet.CLASS_PREFIX + this.convertSelectorToClassName(selector) + ' {' + declarations.join(';') + '}');
			}
			return ruleSets.join('\n');
		},

		serializeColor: function(color) {
			return color && {
				r: color.r,
				g: color.g,
				b: color.b
			};
		},

		serializeStyle: function(style) {
			var font = style.font || {};
			return prelude.Obj.reject(function(arg) {return arg === undefined}, {
				color: this.serializeColor(style.color),
				bgcolor: this.serializeColor(style.background),
				font: font.family,
				height: font.height,
				bold: font.bold,
				italic: font.italic,
				strike: font.strike,
				underline: font.underline
			});
		},

		serializeStyleShort: function(style) {
			var font = style.font || {};
			return prelude.Obj.reject(function(arg) {return arg === undefined}, {
				c: this.serializeColor(style.color),
				bg: this.serializeColor(style.background),
				f: font.family,
				h: font.height,
				b: font.bold,
				i: font.italic,
				s: font.strike,
				u: font.underline
			});
		},

		serialize: function() {
			var styles = {};
			for (var key in this.styles) {
				styles[key] = this.serializeStyle(this.styles[key]);
			}

			return {
				'default': this.serializeStyle(this['default']),
				styles: styles
			};
		}
	}
});





var factory = oop.methodFactory({
	input: {
		properties: [
			{names: ['default', 'def']},
			{names: ['styles'], default: {}}
		]
	},

	process: function(spec) {
		var stylesheet = new Stylesheet();

		// Default -----------------------------------------------------------------

		if (spec.default != null) {
			stylesheet.add(spec.default);
		}

		// Styles ------------------------------------------------------------------

		for (var id in spec.styles) {
			stylesheet.add({
				id: id,
				style: spec.styles[id]
			});
		}

		return stylesheet;
	}
});



exports.Stylesheet = Stylesheet;
exports.factory = factory;
